subprojects/execution/src/main/java/org/gradle/internal/execution/steps/SkipEmptyWorkStep.java

```diff
@@ -22,74 +22,215 @@
 import org.gradle.internal.Try;
 import org.gradle.internal.execution.ExecutionOutcome;
 import org.gradle.internal.execution.ExecutionResult;
+import org.gradle.internal.execution.OutputChangeListener;
 import org.gradle.internal.execution.UnitOfWork;
+import org.gradle.internal.execution.WorkValidationContext;
 import org.gradle.internal.execution.caching.CachingState;
+import org.gradle.internal.execution.fingerprint.InputFingerprinter;
 import org.gradle.internal.execution.history.AfterExecutionState;
+import org.gradle.internal.execution.history.ExecutionHistoryStore;
+import org.gradle.internal.execution.history.OutputsCleaner;
 import org.gradle.internal.execution.history.PreviousExecutionState;
+import org.gradle.internal.fingerprint.CurrentFileCollectionFingerprint;
 import org.gradle.internal.snapshot.FileSystemSnapshot;
+import org.gradle.internal.snapshot.SnapshotUtil;
+import org.gradle.internal.snapshot.ValueSnapshot;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
+import javax.annotation.Nonnull;
+import java.io.File;
+import java.io.IOException;
+import java.io.UncheckedIOException;
 import java.time.Duration;
+import java.util.Map;
 import java.util.Optional;
+import java.util.function.Supplier;
 
-public class SkipEmptyWorkStep<C extends PreviousExecutionContext> implements Step<C, CachingResult> {
-    private final Step<? super C, ? extends CachingResult> delegate;
+public class SkipEmptyWorkStep implements Step<PreviousExecutionContext, CachingResult> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SkipEmptyWorkStep.class);
 
-    public SkipEmptyWorkStep(Step<? super C, ? extends CachingResult> delegate) {
+    private final OutputChangeListener outputChangeListener;
+    private final Supplier<OutputsCleaner> outputsCleanerSupplier;
+    private final Step<? super PreviousExecutionContext, ? extends CachingResult> delegate;
+
+    public SkipEmptyWorkStep(
+        OutputChangeListener outputChangeListener,
+        Supplier<OutputsCleaner> outputsCleanerSupplier,
+        Step<? super PreviousExecutionContext, ? extends CachingResult> delegate
+    ) {
+        this.outputChangeListener = outputChangeListener;
+        this.outputsCleanerSupplier = outputsCleanerSupplier;
         this.delegate = delegate;
     }
 
     @Override
-    public CachingResult execute(UnitOfWork work, C context) {
-        ImmutableSortedMap<String, FileSystemSnapshot> previousOutputFiles = context.getPreviousExecutionState()
+    public CachingResult execute(UnitOfWork work, PreviousExecutionContext context) {
+        ImmutableSortedMap<String, CurrentFileCollectionFingerprint> knownFileFingerprints = context.getInputFileProperties();
+        ImmutableSortedMap<String, ValueSnapshot> knownValueSnapshots = context.getInputProperties();
+        InputFingerprinter.Result newInputs = fingerprintPrimaryInputs(work, context, knownFileFingerprints, knownValueSnapshots);
```

Comment:
I wonder what the best way would be to ensure that all the time spent in snapshotting/fingerpriting is captured in build operations.